### ISSUE TITLE
403 Access Forbidden by CSRFProtector!

### ISSUE THREAD

**Issue by ext10710** [nilsteampassnet/TeamPass#1278]
I made a brand new installation on RHEL 6.6 server, with PHP 5.5.34, Apache 2.2.15 and MySQL 5.1.73 of TeamPass version 2.1.25.2 .  Install went OK, but as soon as I try to change something in the "Settings" tab, I receive this error message:

403 Access Forbidden by CSRFProtector!

Is there an official fix for this, or a workaround?  I browsed the web on this, found that I'm not the only one that experience this behavior, but I didn't find any solution so far...  And this is not an upgrade: as I told before, it's a brand new installation.  Thanks.

**Comment by ext10710** [comment]
Forget it!  Found the answer in this:
https://github.com/nilsteampassnet/TeamPass/issues/1181

Finally, had to replace the "jsUrl" value with:
"jsUrl" => $_SESSION['settings']['cpassman_url']."/includes/libraries/csrfp/js/csrfprotector.js",

in the file "includes/libraries/csrfp/libs/csrfp.config.php".  Works OK now.

**Comment by BradStephens** [comment]
Didn't resolve the issue for me.

**Comment by nilsteampassnet** [comment]
Give a try to branch 2.1.26-RC1

**Comment by BradStephens** [comment]
Still getting the same error when changing settings with that branch.

**Comment by BradStephens** [comment]
I'm guessing it's another issue with MySQL 5.7. Seems that's relatively new.

**Comment by rgalexey** [comment]
Upgraded from 2.1.25.2 to 2.1.26-RC1 and got that issue as well.
- 403 Access Forbidden by CSRFProtector!

**Comment by nilsteampassnet** [comment]
Strange.

Have you cleaned up cache related to teampass in web browser ?
Can i access to server because cannot reproduce?

**Comment by BradStephens** [comment]
I just cleared cookies and disabled cache, still got the error. I can't give access to the server, unfortunately, as it is a company server.

**Comment by rgalexey** [comment]
Unfortunataly no way to give access as well. Since that is corp pass server :)
Not related to browser. I've tried from different PC with fresh browsers.

**Comment by rgalexey** [comment]
I am open to add any debug info on server and test for you if you have any clue.

**Comment by rgalexey** [comment]
And by the way I have 2.1.25.2 on other company server which upgraded recently from 2.1.23.
And there things works fine. 
And teampass/includes/libraries/csrfp/libs/csrfp.config.php  has same right jsUrl like 2.1.26-RC1.

**Comment by nilsteampassnet** [comment]
I really don't understand why you get this error :(
Notice that csrfp was introduced in 2.1.25.

Can you make a test that consists in installing a new instance of 2.1.26-RC1 on a new folder and new database?
Share the content of file */includes/libraries/csrfp/libs/csrfp.config.php*.

**Comment by BradStephens** [comment]
This is from TeamPass 2.1.26-RC1 running on Ubuntu with PHP 7.0.4, mysql 5.7.12.
```
<?php
/**
 * Configuration file for CSRF Protector z
 */

return array(
   "CSRFP_TOKEN" => "50-character-hex-string",
   "logDirectory" => "../log",
   "failedAuthAction" => array(
      "GET" => 0,
      "POST" => 0),
   "errorRedirectionPage" => "",
   "customErrorMessage" => "",
   "jsPath" => "../js/csrfprotector.js",
   "jsUrl" => "http://teampass.company.domain/includes/libraries/csrfp/js/csrfprotector.js",
   "tokenLength" => 25,
   "disabledJavascriptMessage" => "This site attempts to protect users against <a href=\"https://www.owasp.org/index.php/Cross-Site_Request_Forgery_%28CSRF%29\">
   Cross-Site Request Forgeries </a> attacks. In order to do so, you must have JavaScript enabled in your web browser otherwise this site will fail to work correctly for you.
    See details of your web browser for how to enable JavaScript.",
    "verifyGetFor" => array()
);
```

**Comment by nilsteampassnet** [comment]
Perhaps the tokenLength which is wrong. Replace 25 by 50.
The rest is okay

**Comment by BradStephens** [comment]
I think this resolved the issue on my Ubuntu server. I also have a CentOS instance with the issue - here's the csrfp file.

```
<?php
/**
 * Configuration file for CSRF Protector z
 */

return array(
   "CSRFP_TOKEN" => "50-character-hex-string",
   "logDirectory" => "../log",
   "failedAuthAction" => array(
      "GET" => 0,
      "POST" => 0),
   "errorRedirectionPage" => "",
   "customErrorMessage" => "",
   "jsPath" => "../js/csrfprotector.js",
   "jsUrl" => "http://teampass.company.domain/includes/libraries/csrfp/js/csrfprotector.js",
   "tokenLength" => 50,
   "disabledJavascriptMessage" => "This site attempts to protect users against <a href=\"https://www.owasp.org/index.php/Cross-Site_Request_Forgery_%28CSRF%29\">
   Cross-Site Request Forgeries </a> attacks. In order to do so, you must have JavaScript enabled in your web browser otherwise this site will fail to work correctly for you.
    See details of your web browser for how to enable JavaScript.",
    "verifyGetFor" => array()
);
```

**Comment by BradStephens** [comment]
This may also help, from the csrfp log:

```
{
  "timestamp": 1463086681,
  "HOST": "teampass.company.domain",
  "REQUEST_URI": "/index.php?page=manage_settings",
  "requestType": "POST",
  "query": {
    "cpassman_dir": "/var/www/teampass",
    "cpassman_url": "http://teampass.company.domain",
    "path_to_upload_folder": "/var/www/teampass/upload",
    "url_to_upload_folder": "http://teampass.company.domain/upload",
    "path_to_files_folder": "/var/www/teampass/files",
    "url_to_files_folder": "http://teampass.company.domain/files",
    "favicon": "http://teampass.company.domain/favico.ico",
    "custom_logo": "",
    "custom_login_text": "",
    "maintenance_mode_input": "0",
    "default_session_expiration_time": "60",
    "enable_sts_input": "0",
    "encryptClientServer_input": "1",
    "proxy_ip": "",
    "proxy_port": "",
    "pwd_maximum_length": "40",
    "timezone": "UTC",
    "date_format": "d/m/Y",
    "time_format": "H:i:s",
    "default_language": "english",
    "number_of_used_pw": "3",
    "pw_life_duration": "0",
    "nb_bad_authentication": "0",
    "log_connections_input": "0",
    "log_accessed_input": "1",
    "enable_pf_feature_input": "0",
    "use_md5_password_as_salt_input": "0",
    "enable_personal_saltkey_cookie_input": "0",
    "personal_saltkey_cookie_duration": "31",
    "enable_attachment_encryption_input": "0",
    "enable_kb_input": "0",
    "enable_suggestion_input": "0",
    "send_stats_input": "0",
    "get_tp_info_input": "1",
    "delay_item_edition": "0",
    "otv_expiration_period": "7",
    "manager_edit_input": "1",
    "max_last_items": "10",
    "duplicate_folder_input": "0",
    "duplicate_item_input": "0",
    "item_duplicate_in_same_folder_input": "0",
    "show_only_accessible_folders_input": "0",
    "subfolder_rights_as_parent_input": "0",
    "item_extra_fields_input": "0",
    "enable_favourites_input": "1",
    "enable_user_can_create_folders_input": "0",
    "can_create_root_folder_input": "0",
    "activate_expiration_input": "0",
    "enable_delete_after_consultation_input": "0",
    "allow_print_input": "0",
    "roles_allowed_to_print": "0",
    "allow_import_input": "0",
    "anyone_can_modify_input": "0",
    "anyone_can_modify_bydefault_input": "0",
    "restricted_to_input": "0",
    "restricted_to_roles_input": "0",
    "copy_to_clipboard_small_icons_input": "0",
    "show_description_input": "1",
    "tree_counters_input": "0",
    "nb_items_by_query": "auto",
    "enable_send_email_on_user_login_input": "0",
    "enable_email_notification_on_item_shown_input": "0",
    "enable_email_notification_on_user_pw_change_input": "0",
    "insert_manual_entry_item_history_input": "0",
    "settings_offline_mode_input": "0",
    "offline_key_level": "0",
    "syslog_enable_input": "0",
    "syslog_host": "localhost",
    "syslog_port": "514",
    "enable_server_password_change_input": "0",
    "ldap_domain_controler": "0",
    "ldap_ssl_input": "0",
    "ldap_tls_input": "0",
    "ldap_elusers_input": "0",
    "bck_script_path": "/var/www/teampass/backups",
    "bck_script_filename": "bck_cpassman",
    "bck_script_key": "",
    "bck_script_decrypt_file": "",
    "email_smtp_server": "smtp.my_domain.com",
    "email_smtp_auth_input": "0",
    "email_auth_username": "",
    "email_auth_pwd": "",
    "email_server_url": "http://teampass.company.domain",
    "email_port": "25",
    "email_security": "none",
    "email_from": "",
    "email_from_name": "",
    "upload_maxfilesize": "10mb",
    "upload_docext": "doc,docx,dotx,xls,xlsx,xltx,rtf,csv,txt,pdf,ppt,pptx,pot,dotx,xltx",
    "upload_imagesext": "jpg,jpeg,gif,png",
    "upload_pkgext": "7z,rar,tar,zip",
    "upload_otherext": "sql,xml",
    "upload_imageresize_options_input": "1",
    "upload_imageresize_width": "800",
    "upload_imageresize_height": "600",
    "upload_imageresize_quality": "90",
    "save_button": "Save"
  },
  "cookie": {
    "PHPSESSID": "fbj9d8fufdta225s2muvqjdic6",
    "jstree_select": "2",
    "KEY_PHPSESSID": "7h12WezggnAFz8ra2crFSB3NUm+1Woss6FYwRRXx8bI=:BUOSldKiZiyjmOko4I3O2im1zupqCT38DCxlXQGM5Vs=",
    "eaafe88526f2d0454379f521ef84e961065709f99a779f5bd9": "03983c2434b375a8e3de13ce7acc31460812f2dab26c3c83c3"
  }
}
```

**Comment by BradStephens** [comment]
It looks like it didn't resolve the issue on the Ubuntu server either - my mistake.

**Comment by nilsteampassnet** [comment]
Yes I experimented the same by using another test environment.

It is strange. In my case, it doesn't occurs when using WampServer (with PHP7), but occurs when using Xampp (with PHP5).

I'm currently by-passing this (as I don't understand the reason) by changing the way Settings page will store the settings. In 2.1.27, the save is done directly when you perform a change, and no "page submit" is performed any more.

**Comment by BradStephens** [comment]
This is working correctly on a fresh installation of the 2.1.26 (2) release.

**Comment by aostrovsky** [comment]
Confirmed working 2.1.26.2

**Comment by duss69** [comment]
Hi !!

I have same issue on Teampass 2.1.26 (distrib centos 7 ) with Https. 


This my /var/www/html/fr-teampassit/includes/libraries/csrfp/libs/csrfp.config.php
"
<?php
/**
 * Configuration file for CSRF Protector z
 */

return array(
   "CSRFP_TOKEN" => "_**bb2641249928ea14d5795e17ddd7b551016f1600c798e74623**_",
   "logDirectory" => "../log",
   "failedAuthAction" => array(
      "GET" => 0,
      "POST" => 0),
   "errorRedirectionPage" => "",
   "customErrorMessage" => "",
   "jsPath" => "../js/csrfprotector.js",
   "jsUrl" => "http://10.33.34.60/teampass/includes/libraries/csrfp/js/csrfprotector.js",
   "tokenLength" => 50,
   "disabledJavascriptMessage" => "This site attempts to protect users against <a href=\"https://www.owasp.org/index.php/Cross-Site_Request_Forgery_%28CSRF%29\"
>
   Cross-Site Request Forgeries </a> attacks. In order to do so, you must have JavaScript enabled in your web browser otherwise this site will fail to work corr
ectly for you.
    See details of your web browser for how to enable JavaScript.",
    "verifyGetFor" => array()
);


Is it normal that my  CSRFP_TOKEN is not like "50-character-hex-string",

Thank .

**Comment by cgkelleher** [comment]
Hi, 

I am also receiving this error message although only when trying to restore a backup.

Strange thing is that the backup is successful and restores the folders i deleted for testing?

thanks

**Comment by jtbw911** [comment]
I am experiencing this on the latest pull - 2.1.27. I updated the Teampass URLs to HTTPS, and configured a redirect on Apache, and it started doing so. I tried restarting the service, the server, the clients, cleared cookies, etc. and then even reverted the changes to no avail.

`<?php /** * Configuration file for CSRF Protector z */ return array( "CSRFP_TOKEN" => "50-character token", "logDirectory" => "../log", "failedAuthAction" => array( "GET" => 0, "POST" => 0), "errorRedirecti> "", "customErrorMessage" => "", "jsPath" => "../js/csrfprotector.js", "jsUrl" => $_SESSION['settings']['cpassman_url']."/includes/libraries/csrfp/js/csrfprotector.js", "tokenLength" => "50", "disabledJavascriptMessage" => "This site attempts to protect users against Cross-Site Request Forgeries attacks. In order to do so, you must have JavaScript enabled in your web browser otherwise this site will fail to work correctly for you. See details of your web browser for how to enable JavaScript.", "verifyGetFor" => array("*type=duo_check*", "*upload.attachments.php*", "*upload.files.php*") );`

I had also tried it this way as well:

`&lt;?php
/**
 * Configuration file for CSRF Protector z
 */

return array(
    "CSRFP_TOKEN" => "50-character token",
    "logDirectory" => "../log",
    "failedAuthAction" => array(
        "GET" => 0,
        "POST" => 0),
    "errorRedirecti> "",
    "customErrorMessage" => "",
    "jsPath" => "../js/csrfprotector.js",
    "jsUrl" => "http:/teampass.company.com/includes/libraries/csrfp/js/csrfprotector.js",
    "tokenLength" => "50",
    "disabledJavascriptMessage" => "This site attempts to protect users against <a >
   Cross-Site Request Forgeries </a> attacks. In order to do so, you must have JavaScript enabled in your web browser otherwise this site will fail to work correctly for you.
    See details of your web browser for how to enable JavaScript.",
    "verifyGetFor" => array("*type=duo_check*", "*upload.attachments.php*", "*upload.files.php*")
);`

**Comment by nilsteampassnet** [comment]
You write that you have done a redirection on apache. Please don't do that and set up à virtualhost on port 443.
In your tp.config.php, please be sure to keep the * in   the verifygetfor and jsUrl must have the full url.

**Comment by jtbw911** [comment]
My apologies. Let me rephrase that. We are doing a permanent redirection on a virtualhost:80 to https://teampass.company.com that runs as teampass.company.com on virtualhost:443.

I have reverted the csfrp.config.php to this:

`<?php /** * Configuration file for CSRF Protector z */ return array( "CSRFP_TOKEN" => "50-character token", "logDirectory" => "../log", "failedAuthAction" => array( "GET" => 0, "POST" => 0), "errorRedirecti> "", "customErrorMessage" => "", "jsPath" => "../js/csrfprotector.js", "jsUrl" => "https://teampass.acendas.com/includes/libraries/csrfp/js/csrfprotector.js", "tokenLength" => "50", "disabledJavascriptMessage" => "This site attempts to protect users against Cross-Site Request Forgeries attacks. In order to do so, you must have JavaScript enabled in your web browser otherwise this site will fail to work correctly for you. See details of your web browser for how to enable JavaScript.", "verifyGetFor" => array("*type=duo_check*", "*upload.attachments.php*", "*upload.files.php*") );`

I still receive the error upon just browsing to the site (not even able to login to get to settings). I have removed the permanent redirect virtualhost:80 to no avail.

I would be more than happy to setup a time for you to take a look at the site.

**Comment by jtbw911** [comment]
I was able to resolve it by uploading the /install folder, running the installation over top of the existing installation, and configuring it again. This seems to have preserved the data from the database while resolving the csrfp issue.

**Comment by silasfn** [comment]
I got this error on a fresh install, and was able to resolve by fixing the permissions on
/var/lib/php/7.0/session/
```
sudo chown -R root:nginx /var/lib/php/7.0/*
```
I found the issue by checking /var/log/nginx/error.log

**Comment by niedlipatrik** [comment]
I had the same problem, I managed to solve it by creating a folder with editing permission in a different location on the example server: "public_html/edoc21/odm_data" and put this path in the database in the table "odm_settings" field "dataDir"
Now it is working, by default the installation creates a folder to which we do not have access and even if we point to the installation it seems that it does not write to the database.
